design-proposals/klip-57-kafka-headers.md

```diff
@@ -0,0 +1,141 @@
+* Including header columns in a data source's schema
+* Functions to help decode byte strings to primitive types (BYTES to STRING already exists, new functions would be conversions to INT, BIGINT and DOUBLE)
+
+## What is not in scope
+* Writing headers to output topics - we might do this in the future, but that is a separate discussion
+* Advanced functions to help decode byte strings (e.g. functions to deserialize complex serialized data.) - if a user wants to convert something that we don't have a function for, then they can either create their own UDF or use `CREATE FUNCTION` when that becomes available.
+
+## Public APIS
+
+### Syntax
+
+Unlike other pseudocolumns, if a user wants to access headers, they will create a column with the `HEADERS` keyword. This column will be populated with the full list of header keys and values. There can only be one HEADERS column.
+```
+ksql> CREATE STREAM A (
+    my_headers ARRAY<STRUCT<key STRING, value BYTES>> HEADERS);
+```
+If there are multiple HEADERS columns, then an error will be thrown. The type of a header backed column must be exactly `ARRAY<STRUCT<key STRING, value BYTES>>`. An error will be thrown if that is not the case.
+```
+ksql> CREATE STREAM B (my_headers ARRAY<BYTES> HEADERS);
+Error: Columns specified with the HEADERS keyword must be typed as ARRAY<STRUCT<key STRING, value BYTES>>.
+```
+Users can also mark a column with `HEADER(<key>)`, which will populate the column with the last header that matches the key. The data type must be `BYTES`
+```
+ksql> CREATE STREAM B (value BYTES HEADER(<key>), value_2 BYTES HEADER(<key_2>));
+```
+If there is a column labelled `HEADERS`, then no there can be no `HEADER(<key>)` columns. Otherwise, there can be multiple `HEADER(<key>)` columns, but no two can share the same key. If the key is not present in the header, then the value in the column will be `null`.
+
+Sources created with schema inferencing can have header columns, they would just have to be specified. For example, the statement `CREATE STREAM A (my_headers ARRAY<STRUCT<key STRING, value BYTES>> HEADERS) WITH (kafka_topic='a', value_format='avro');` will create a stream with columns from Schema Registry, as well as one header column.
+
+### New byte-conversion functions
+
+There will also be new byte conversion functions to help decode header data:
+
+* `INT_FROM_BYTES(bytes, byte_order='big', signed=true)` - The input byte array must be exactly 4 bytes long. Signed conversions will by handled by `ByteBuffer`'s `getInt` method. Unsigned conversions will be manually calculated. If the result does not fit into an INTEGER then the method will throw an error.
+* `BIGINT_FROM_BYTES(bytes, byte_order='big', signed=true)` - The input byte array must be exactly 8 bytes long. Signed conversions will by handled by `ByteBuffer`'s `getLong` method. Unsigned conversions will be manually calculated. If the result does not fit into a BIGINT then the method will throw an error.
```

Comment:
ksql does not support unsigned integers, so I don't think we can use the `signed` parameter. In java, for instance, unsigned integers are stored in long values in order to fit the unsigned value there.

Is this `signed` a requirement?

A similar question about the byte order. Seems any java API to convert to integer uses big endian. Are we expecting users to use different orders than just using the API to convert from int -> byte array?

Reply:
Users are limited in what they are able to convert their bytes to. The problem with signed/unsigned is that the byte representations of the two are completely different, so if we only support signed, then a user who wants to do an unsigned conversion would have no way to do it. We could get around this by adding byte arithmetic, but that would still be a big headache to do.

Reply:
The ksql `INT` type cannot hold an unsigned integer. When ksql manages `INT` types, it uses the `java.lang.Integer` which holds only `-214748364` to `2147483647`. An unsigned integer should hold `0` to `4294967295`, which is not what `INT` can support.

The option should be to support `INT UNSIGNED` types, which is a bigger scope.

Reply:
The `If the result does not fit into an INTEGER then the method will throw an error` is supposed to cover that. I can see how it's awkward.

@colinhicks are `UNSIGNED` numbers or bit manipulation on the roadmap? Either of those would give users a workaround.

As for the byte order, it's easy for us to flip a byte array and some applications do use little endian, so I think we should keep it.